src/xmagics.hpp

```diff
@@ -0,0 +1,251 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Loic Gouarin and Sylvain Corlay       *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+#ifndef XMAGICS_HPP
+#define XMAGICS_HPP
+
+#include <algorithm>
+#include <cstdlib>
+#include <fstream>
+#include <memory>
+#include <regex>
+#include <type_traits>
+
+#include "cling/Interpreter/Value.h"
+#include "cling/Interpreter/Exception.h"
+#include "cling/Utils/Output.h"
+
+#include "xparser.hpp"
+
+namespace xeus
+{
+    struct XEUS_API xmagic_line
+    {
+        using type = xmagic_line;
+        virtual void operator()(const std::string& line) const{};
+    };
+    
+    struct XEUS_API xmagic_cell
+    {
+        virtual void operator()(const std::string& line, const std::string& cell) const{};
+    };
+
+    struct XEUS_API xmagic_line_cell: public xmagic_line, xmagic_cell
+    {
+    };
+
+    class XEUS_API xmagics_manager 
+    {
+    public:
+
+        xmagics_manager(){};
+
+        template<typename xmagic_type>
+        void register_magic(const std::string& magic_name, std::shared_ptr<xmagic_type> magic)
```

Comment:
I think we could use `std::unique_ptr` here. I will fix that and I will define an alias.